example/hello.cc

```diff
@@ -1,93 +1,80 @@
-#include <iostream>
+#include <cstddef>
 #include <fstream>
-#include <cstdlib>
-#include <string>
-#include <cinttypes>
-
-#include "wasm.hh"
+#include <functional>
+#include <iostream>
+#include <utility>
+#include <vector>
 
+#include "imaginary-wasm.hh"
 
-// A function to be called from Wasm code.
-auto hello_callback(
-  const wasm::vec<wasm::Val>& args, wasm::vec<wasm::Val>& results
-) -> wasm::own<wasm::Trap> {
-  std::cout << "Calling back..." << std::endl;
-  std::cout << "> Hello world!" << std::endl;
-  return nullptr;
+namespace {
+std::vector<char> load_file(const std::string &filename) {
+  std::ifstream file(filename, std::ios::binary | std::ios::ate);
+  auto file_size = file.tellg();
+  std::vector<char> buffer(file_size);
+  file.seekg(0);
+  if (file.fail() || !file.read(buffer.data(), file_size)) {
+    std::cout << "> Error loading module!" << std::endl;
+    exit(1);
+  }
+  return buffer;
 }
-
+} // namespace
 
 void run() {
   // Initialize.
   std::cout << "Initializing..." << std::endl;
-  auto engine = wasm::Engine::make();
-  auto store_ = wasm::Store::make(engine.get());
-  auto store = store_.get();
+  wasm::Store store;
 
   // Load binary.
   std::cout << "Loading binary..." << std::endl;
-  std::ifstream file("hello.wasm");
-  file.seekg(0, std::ios_base::end);
-  auto file_size = file.tellg();
-  file.seekg(0);
-  auto binary = wasm::vec<byte_t>::make_uninitialized(file_size);
-  file.read(binary.get(), file_size);
-  file.close();
-  if (file.fail()) {
-    std::cout << "> Error loading module!" << std::endl;
-    exit(1);
-  }
-
-  // Compile.
-  std::cout << "Compiling module..." << std::endl;
-  auto module = wasm::Module::make(store, binary);
-  if (!module) {
-    std::cout << "> Error compiling module!" << std::endl;
-    exit(1);
-  }
+  auto file = load_file("hello.wasm");
 
-  // Create external print functions.
-  std::cout << "Creating callback..." << std::endl;
-  auto hello_type = wasm::FuncType::make(
-    wasm::ownvec<wasm::ValType>::make(), wasm::ownvec<wasm::ValType>::make()
-  );
-  auto hello_func = wasm::Func::make(store, hello_type.get(), hello_callback);
+  // Here instantiation is performed straight from .wasm bytes. You can compile
+  // and instantiate in two steps with:
+  //   wasm::Module module = store.compile(file);
+  //   if (auto error = module.get_error()) { /* ... */ }
+  //   wasm::Instance instance = module.instantiate(wasm::Imports {});
 
   // Instantiate.
   std::cout << "Instantiating module..." << std::endl;
-  auto imports = wasm::vec<wasm::Extern*>::make(hello_func.get());
-  auto instance = wasm::Instance::make(store, module.get(), imports);
-  if (!instance) {
-    std::cout << "> Error instantiating module!" << std::endl;
+  wasm::Imports imports{
+    {
+      "",
+      {
+        {"hello", [] { std::cout << "> Hello world!" << std::endl; }}
+      },
+    }
+  };
+
+  wasm::Instance instance = store.instantiate(file, imports);
```

Comment:
Unsure about this being a method on `store`, but it's not really my place to say. If you think this is better, then we should get feedback from the Wasm C API maintainers about it!

Reply:
It's just a convenience function. It's equivalent to:
```c++
wasm::Module module = store.compile(file);
wasm::Instance instance = module.instantiate(imports);
```
or
```c
own wasm_module_t* module = wasm_module_new(store, &binary);
own wasm_instance_t* instance =
  wasm_instance_new(store, module, &imports, NULL);
```

Reply:
I think we should keep the "standard" C++ API free from convenience method and ship them through a wrapper of our own or something like that.

Reply:
I think it might be hard to add that on after the fact. Notably the standard Wasm C API does have some convenience functions included in `wasm.h`... as long as the convenience method doesn't make it any harder on the implementors of the API (i.e. it's all done in the header file), I think it's probably fine to include some within reason.

I'm not completely convinced that the `store.` pattern is the one we'd want, but I haven't thought it through all the way. I see the motivation behind it... `store` is the first argument to a looot of functions.